Export scss vars as json during compilation

weltenwort opened the pull request:
This PR enhances the script `scripts/compile-scss.js` to export also extract and export the sass variables from each theme into a corresponding json file in `dist/eui_theme_${THEME}.json`. The values in this file are in a format suitable for use in styled-components themes to keep the scss files as a single source of truth.

It also converts the script to use async I/O to improve parallelism.

The export of ready-to-use themes could be considered in a subsequent step.

cjcenizal commented:
@weltenwort Thanks for doing this! I won't be able to review this so I'm tagging in @mattapperson and @snide.

sqren commented on scripts/compile-scss.js:
```diff
@@ -1,16 +1,53 @@
+      const splitFileName = fileName.split('.');
+      const baseFileName = splitFileName[0];
+      const cssFileName = `dist/eui_${baseFileName}.css`;
+      const varsFileName = `dist/eui_${baseFileName}.json`;
```
I know you didn't write this code, but since we are at it, perhaps replace all filename logic with [path.parse](https://nodejs.org/api/path.html#path_path_parse_path). You can get it mostly with a one-liner, and it'll work cross platform (I don't think Window likes the forward slashes):
```js
const { name: baseFileName } = path.parse(file);
```

sqren commented on scripts/compile-scss.js:
```diff
@@ -1,16 +1,53 @@
+        });
+    })
+  )
+);
```
Maybe a `catch` at the bottom, so errors doesn't get swallowed?

sqren replied:
nit: I think this would benefit from async/await.

sqren commented:
This is really great. Thanks for doing this!! 👍

weltenwort commented:
Thanks for the very valuable feedback, @sqren!

I have flattened the `Promise` chains a bit using aync/await and split it up into several functions. (Too bad that node 8 does not yet support `for await ... of`.) I also removed the superfluous `shelljs` dependency and try to handle all paths in a cross-platform manner using `path.parse` and `path.join`. Additionally, the script now isolates errors while processing each file.

Let me know what you think.

mattapperson commented:
@weltenwort I am curious if it is better to save as a JSON file or a js file exporting an object. JSON will have parsing overhead at runtime as webpack will bundle as a string then JSON.parse on each require (at least last time I checked)

mattapperson commented:
@weltenwort Well, yea writing to SRC is bad. So what if we then create an export on EUI for the JSON file that pre-parses it one time, preventing multiple parsing?

mattapperson commented:
Another added perk to exporting the configs is it gives it an API rather then just a file. This way we can change the implementation without breaking compatibility if we need to down the road

weltenwort commented:
I agree that exposing a javascript interface would be nice, but...

That adds the assumption that any consumer of the commonjs or es module can import json. AFAIK that is the case for node, webpack and browserify, but I don't know if that's sufficient.

We *could* put the json import in something like `src/themes/index.js` without re-exporting it in `src/index.js`, so it's up to the consumer to execute that code path. :thinking:

weltenwort commented:
thanks for sharing your thoughts, @snide

**breaking changes**: That is already the case with just `scss`. Renaming variables and mixins would break any `scss` code that imports it, even when it's statically compiled. If we say "please re-use the variables and mixins when writing new styles", we make them part of the public api.

**mixins**: Somehow "importing" the `scss` mixins will be difficult since they can contain complex logic. A solution would probably be so complicated that I would question the benefits of using it.

Looking at `css-literal-loader`, it seems that its approach is to extract the style snippets from js files and statically compile them at build time. That is different from what most css-in-js solutions do, because it prevents most kinds of dynamic style changes.

So the question seems to be whether we can live without reusing the mixins in components outside of EUI. @sqren, having used styled-components in production while trying to match EUI styling, did you miss the ability to access the sass mixins so far?

sqren commented:
> So the question seems to be whether we can live without reusing the mixins in components outside of EUI. @sqren, having used styled-components in production while trying to match EUI styling, did you miss the ability to access the sass mixins so far?

I didn't miss sass mixins, since mixins feel very natural in js (example from the styled-component docs):
```js
import { truncate } from '../style-utils';

// Make this div truncate the text with an ellipsis
const Box = styled.div`
  ${ truncate('250px') }
  background: papayawhip;
`;
```

If a lot of work has been put into creating sass mixins, it will of course take time to redo that in js. But until now I haven't really missed EUI mixins. I mostly use EUI's react components and configure them through props. 
I use styled-components for custom details, and minor things that EUI doesn't have to care about.

mattapperson commented:
@felix-lessoer I am fine with adding it to `src/themes/index.js`

alternatively though, what if we add something like the solution outlined here https://github.com/webpack/webpack/issues/6572. That way the JSON is just included in the bundle as an object?

weltenwort commented:
I would like to avoid cluttering the build process with more and more loaders... especially since the "bundle" is only one of three ways to consume EUI. Kibana is probably using the es-module variant straight from `src`.

(also, wrong Felix :wink:)

mattapperson commented:
But it’s not adding the *need* for another loader. If you have the loader you get better perf then just loading the json and parsing at runtime... graceful degradation. We can still put the access to the api inside arc/themes/index.js or whatever you like for that

mattapperson commented:
WOOT! Thanks for this @weltenwort!
